Render the index/default page completely client-side

eirslett opened the pull request:
> The default page used to leverage a /modelview endpoint
> on the Scala server to generate a model to pass into its
> Mustache template. The logic from that /modelview endpoint
> is now ported to JavaScript, so that all the pre-rendering
> business logic happens on the client side.
> 
> Since this is a high-risk change, the /modelview endpoint
> still exists on the server. Every time we request a page,
> the UI will actually fetch a processed version from that
> endpoint as well, and compare the Scala version with
> the JavaScript version. The idea is that we can compare
> outputs of the two, and if there's a difference, we can
> assume that's a bug in the JavaScript port.
> The endpoint, and the UI logic to compare implementations,
> will soon be removed, after testing.

Please don't merge this PR yet - I really need people to test
this change manually!

It would be great if you can checkout this branch, start
zipkin-web and look at a few trace summaries. (Not traces,
but the list of trace summaries/search results on the front
page).
In the browser debugger, there's a console.log thing that
logs whether the JavaScript implementation gives the same
result as the Scala implementation, given the same input.

I would love to hear if anybody experiences a situation where
the two implementations produce different output - and in
that case, what the input was.

If you've tried running zipkin-web with this change, and you
found no bugs, be sure to +1 the PR. Whenever we're sure
that most bugs are cleared out, I'll update with another commit
that removes the automatic scala-vs-javascript-comparison
on every page load.

eirslett committed:
Render the index/default page completely client-side

The default page used to leverage a /modelview endpoint
on the Scala server to generate a model to pass into its
Mustache template. The logic from that /modelview endpoint
is now ported to JavaScript, so that all the pre-rendering
business logic happens on the client side.

Since this is a high-risk change, the /modelview endpoint
still exists on the server. Every time we request a page,
the UI will actually fetch a processed version from that
endpoint as well, and compare the Scala version with
the JavaScript version. The idea is that we can compare
outputs of the two, and if there's a difference, we can
assume that's a bug in the JavaScript port.
The endpoint, and the UI logic to compare implementations,
will soon be removed, after testing.

eirslett commented on zipkin-web/package.json:
```diff
@@ -20,7 +20,8 @@
     "js-cookie": "^2.1.0",
-    "moment": "^2.11.1",
+    "lodash": "^4.5.0",
+    "moment": "^2.11.2",
```
lodash is great for doing functional stuff on collections, mimicking the Scala implementation which uses Scala's collection API a lot. Moment is for formatting datetime.

codefromthecrypt replied:
great. I know we had an outstanding issue wrt moment.. particularly the duration thing, which defaults to user hard-entering micros, should be easier now! #864 cc @danchia @yurishkuro

eirslett commented on zipkin-web/src/main/resources/app/js/component_data/default.js:
```diff
@@ -1,15 +1,41 @@
 import {component} from 'flight';
 import $ from 'jquery';
+import queryString from 'query-string';
+import chai from 'chai';
```
chai is an assertion framework used for javascript unit tests, we'd never run it in production code but I've included it here temporarily to perform the deep-equals comparison.

codefromthecrypt replied:
maybe convert these two comments to code comments, as they are less likely to get lost.

eirslett replied:
That's the whole point, it's going to get lost. (Chai will be removed as a dependency, only kept in tests.)
Commenting about Chai in the unit tests is somewhat redundant, since it's a core part of the node.js ecosystem. It's like explaining in src/test/scala what ScalaTest is.

codefromthecrypt replied:
oh ok nevermind then

eirslett commented on zipkin-web/src/main/resources/app/js/component_data/default.js:
```diff
@@ -1,15 +1,41 @@
+        success: (traces) => {
+          const modelview = {traces: traceSummariesToMustache(serviceName, traces.map(traceSummary)) };
+          this.trigger('defaultPageModelView', modelview);
+          if (serviceName) {
```
I guess this if-check is redundant, but it will be removed anyways, when we remove the comparison code.

codefromthecrypt replied:
ack

eirslett commented on zipkin-web/src/main/scala/com/twitter/zipkin/web/Handlers.scala:
```diff
@@ -178,7 +178,7 @@ class Handlers(queryExtractor: QueryExtractor) {
       val serviceDurations = groupedSpanTimestamps.map { case (n, sts) =>
         MustacheServiceDuration(n, sts.length, sts.map(_.duration).max / 1000)
-      }.toSeq
+      }.toSeq.sortWith((d1, d2) => d1.name.compareTo(d2.name) < 0)
```
Sorting service durations alphabetically by name was more deterministic than the existing sort order (which seems random)

codefromthecrypt replied:
agreed

eirslett commented on zipkin-web/src/main/scala/com/twitter/zipkin/web/Handlers.scala:
```diff
@@ -214,9 +214,7 @@ class Handlers(queryExtractor: QueryExtractor) {
         val data = Map(
-          "traces" -> traceSummaryToMustache(serviceName, traces),
-          "annotations" -> annotations,
-          "binaryAnnotations" -> binaryAnnotations
```
It didn't look like annotations and binaryAnnotations were in use, in index.mustache.

codefromthecrypt replied:
I was curious about that, too

codefromthecrypt commented on zipkin-web/src/main/resources/app/test/component_ui/traceSummary.test.js:
```diff
@@ -0,0 +1,235 @@
+import {
```
wat? a test class?! sorry I must be in the wrong project...

eirslett replied:
:stuck_out_tongue_winking_eye:

codefromthecrypt commented:
looks like a mild precision error in the assertion of durationStr.

```
AssertionError: expected [ { traceId: '56bea38b40e95f83',
    startTs: '02-29-2016T09:09:45.388+0800',
    timestamp: 1456708185388000,
    duration: 3948.889,
    durationStr: '3948.889ms',
    servicePercentage: 100,
    spanCount: 1,
    serviceDurations: [ [Object], remove: [Function] ],
    width: 100 },
  remove: [Function] ] to deeply equal [ { traceId: '56bea38b40e95f83',
    startTs: '02-29-2016T09:09:45.388+0800',
    timestamp: 1456708185388000,
    duration: 3948,
    durationStr: '3.948s',
    servicePercentage: 100,
    spanCount: 1,
    serviceDurations: [ [Object], remove: [Function] ],
    width: 100 },
  remove: [Function] ]
```

codefromthecrypt commented:
Here's a longer one..

```
Uncaught AssertionError: expected [ { traceId: '56bea38b40e95f83',
    startTs: '02-29-2016T09:09:45.388+0800',
    timestamp: 1456708185388000,
    duration: 3948.889,
    durationStr: '3948.889ms',
    servicePercentage: 100,
    spanCount: 1,
    serviceDurations: [ [Object], remove: [Function] ],
    width: 100 },
  { traceId: '249628f4fa8f23d3',
    startTs: '02-29-2016T09:21:10.066+0800',
    timestamp: 1456708870066000,
    duration: 50,
    durationStr: '50.000ms',
    servicePercentage: 57,
    spanCount: 8,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 1 },
  { traceId: '6d59dbe7ea141366',
    startTs: '02-29-2016T09:21:10.066+0800',
    timestamp: 1456708870066000,
    duration: 50,
    durationStr: '50.000ms',
    servicePercentage: 57,
    spanCount: 8,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 1 },
  { traceId: 'ef2ce5f890aee2dd',
    startTs: '02-29-2016T09:21:21.907+0800',
    timestamp: 1456708881907000,
    duration: 16,
    durationStr: '16.000ms',
    servicePercentage: 87,
    spanCount: 8,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 0 },
  { traceId: '57b9953167173b3f',
    startTs: '02-29-2016T09:22:56.985+0800',
    timestamp: 1456708976985000,
    duration: 13,
    durationStr: '13.000ms',
    servicePercentage: 100,
    spanCount: 5,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 0 },
  { traceId: 'fe33e75d70447b1d',
    startTs: '02-29-2016T09:21:21.966+0800',
    timestamp: 1456708881966000,
    duration: 6,
    durationStr: '6.000ms',
    servicePercentage: 66,
    spanCount: 8,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 0 },
  { traceId: '7318dcb23457300f',
    startTs: '02-29-2016T09:21:21.966+0800',
    timestamp: 1456708881966000,
    duration: 6,
    durationStr: '6.000ms',
    servicePercentage: 66,
    spanCount: 8,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 0 },
  { traceId: '2a47be612f5c86b7',
    startTs: '02-29-2016T09:21:22.020+0800',
    timestamp: 1456708882020000,
    duration: 3,
    durationStr: '3.000ms',
    servicePercentage: 100,
    spanCount: 5,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 0 },
  remove: [Function] ] to deeply equal [ { traceId: '56bea38b40e95f83',
    startTs: '02-29-2016T09:09:45.388+0800',
    timestamp: 1456708185388000,
    duration: 3948,
    durationStr: '3.948s',
    servicePercentage: 100,
    spanCount: 1,
    serviceDurations: [ [Object], remove: [Function] ],
    width: 100 },
  { traceId: '6d59dbe7ea141366',
    startTs: '02-29-2016T09:21:10.066+0800',
    timestamp: 1456708870066000,
    duration: 50,
    durationStr: '50.000ms',
    servicePercentage: 58,
    spanCount: 4,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 1 },
  { traceId: '249628f4fa8f23d3',
    startTs: '02-29-2016T09:21:10.066+0800',
    timestamp: 1456708870066000,
    duration: 50,
    durationStr: '50.000ms',
    servicePercentage: 58,
    spanCount: 4,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 1 },
  { traceId: 'ef2ce5f890aee2dd',
    startTs: '02-29-2016T09:21:21.907+0800',
    timestamp: 1456708881907000,
    duration: 16,
    durationStr: '16.000ms',
    servicePercentage: 87,
    spanCount: 4,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 0 },
  { traceId: '57b9953167173b3f',
    startTs: '02-29-2016T09:22:56.985+0800',
    timestamp: 1456708976985000,
    duration: 13,
    durationStr: '13.000ms',
    servicePercentage: 100,
    spanCount: 3,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 0 },
  { traceId: '7318dcb23457300f',
    startTs: '02-29-2016T09:21:21.966+0800',
    timestamp: 1456708881966000,
    duration: 6,
    durationStr: '6.000ms',
    servicePercentage: 66,
    spanCount: 4,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 0 },
  { traceId: 'fe33e75d70447b1d',
    startTs: '02-29-2016T09:21:21.966+0800',
    timestamp: 1456708881966000,
    duration: 6,
    durationStr: '6.000ms',
    servicePercentage: 66,
    spanCount: 4,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 0 },
  { traceId: '2a47be612f5c86b7',
    startTs: '02-29-2016T09:21:22.020+0800',
    timestamp: 1456708882020000,
    duration: 3,
    durationStr: '3.000ms',
    servicePercentage: 100,
    spanCount: 3,
    serviceDurations: [ [Object], [Object], remove: [Function] ],
    width: 0 },
  remove: [Function] ]
```

Here's the associated json:

``` json
[{"traceId":"56bea38b40e95f83","name":"bootstrap","id":"56bea38b40e95f83","timestamp":1456708185388000,"duration":3948889,"annotations":[{"endpoint":{"serviceName":"zipkin-query","ipv4":"192.168.99.1","port":9411},"timestamp":1456708185437062,"value":"ApplicationStarted"},{"endpoint":{"serviceName":"zipkin-query","ipv4":"192.168.99.1","port":9411},"timestamp":1456708185615720,"value":"ApplicationEnvironmentPrepared"},{"endpoint":{"serviceName":"zipkin-query","ipv4":"192.168.99.1","port":9411},"timestamp":1456708185720714,"value":"ApplicationPrepared"},{"endpoint":{"serviceName":"zipkin-query","ipv4":"192.168.99.1","port":9411},"timestamp":1456708189236645,"value":"ContextRefreshed"},{"endpoint":{"serviceName":"zipkin-query","ipv4":"192.168.99.1","port":9411},"timestamp":1456708189332465,"value":"EmbeddedServletContainerInitialized"},{"endpoint":{"serviceName":"zipkin-query","ipv4":"192.168.99.1","port":9411},"timestamp":1456708189336886,"value":"ApplicationReady"}],"binaryAnnotations":[{"key":"lc","value":"spring-boot","endpoint":{"serviceName":"zipkin-query","ipv4":"192.168.99.1","port":9411}}],"debug":false}]
```

You can post the json to your query server like so.. `curl -s localhost:9411/api/v1/spans -X POST -H "Content-Type: application/json" -d ...`

eirslett commented:
Thanks a lot! :-D I'll have a look at the diffs!
durationStr is somewhat dumber implemented in JavaScript, it only shows it as number of milliseconds with 3 digits precision. I'll have to implement the Scala logic that shows the best measuring unit based on the given value.

eirslett committed:
Client-side rendering of traces

This is a JavaScript port of all the Scala business logic for
rendering a trace with Mustache. Now, everything happens
client-side, based on API-calls to zipkin-query.

The same JS-vs-Scala comparison functionality as the one from
the default page is implemented here, to try and detect
regressions by comparing real-life data. This comparison
will be removed before merging to the master branch.

mkDurationStr was updated with a few minor fixes to make it
resemble the original more closely.
shallow-deep-almost-equal is used for the comparison, so that
it assumes 1.0000001 == 1.000000 (To take into account
JavaScript rounding errors)

eirslett commented on zipkin-web/src/main/scala/com/twitter/zipkin/web/Handlers.scala:
```diff
@@ -15,6 +15,8 @@ import com.twitter.conversions.time._
 import org.jboss.netty.handler.codec.http.QueryStringEncoder
 import java.io.InputStream
 
+import sun.net.www.http.HttpClient
```
:-/ not sure how this ended up here, will remove it before merge

codefromthecrypt replied:
this broke my build, btw

eirslett replied:
Should be fixed now. I updated the PR.

eirslett commented on zipkin-web/src/main/scala/com/twitter/zipkin/web/Handlers.scala:
```diff
@@ -293,7 +295,7 @@ class Handlers(queryExtractor: QueryExtractor) {
         "depth" -> (depth + 1) * 5,
         "depthClass" -> (depth - 1) % 6,
-        "children" -> spanMap.get(span.id).map(s => SpanId(s.id).toString).mkString(","),
+        "children" -> byParentId.getOrElse(span.id, Nil).map(s => SpanId(s.id).toString).mkString(","),
```
This is basicalle #1014, which was needed here, since it's implemented that way in JS.

eirslett commented:
I've now pushed a similar JavaScript port of the trace rendering.
It's being compared in the browser the same way as with the index/default page.
More unit tests would be nice, but I'm also looking forward to receiving some feedback from people who have tested this with real-life traces? @adriancole @yurishkuro do you have time to run this branch locally and look for errors in the browser console debugger?

When this PR is fixed and merged, all the heavy lifting is done. There's no more business logic on the server side of zipkin-web (it's basically a reverse proxy + static file server, written in Scala), so we will almost be ready to package zipkin-web as a .zip/tar.gz of assets, and possibly serve them directly from zipkin-query.

eirslett commented:
@adriancole the durationStr thing should be fixed now. Was that the only problem with the JSON you posted?

eirslett committed:
remove unneeded import

eirslett committed:
remove duplicate declaration in javascript test

codefromthecrypt commented:
only found two last glitches:

```
app.min.js:19480 Uncaught AssertionError: Expected to have "127.0.0.1:0" but got "127.0.0.1:undefined" at path "/spansBackup/0/annotations/0/endpoint".
```

And this..

```
app.min.js:19480 Uncaught AssertionError: Expected to have "1+-0.001" but got "3" at path "/0/spanCount".
```

Above for the following trace:

``` json
[{"traceId":"d397ce70f5192a8b","name":"get","id":"d397ce70f5192a8b","timestamp":1457160374149000,"duration":2000,"annotations":[{"timestamp":1457160374149000,"value":"sr","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}},{"timestamp":1457160374151000,"value":"ss","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}}],"binaryAnnotations":[{"key":"http.path","value":"/api/v1/services","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}},{"key":"srv/finagle.version","value":"6.33.0","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}},{"key":"sa","value":true,"endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}},{"key":"ca","value":true,"endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":56828}}]}]
```

eirslett committed:
Format endpoint: default port should be 0

Also add unit tests to confirm the functionality.

eirslett committed:
fix error that showed wrong span count

eirslett commented:
Thanks for reporting! The two errors should both be fixed now.

codefromthecrypt commented:
Thanks for your patience, Eirik. Found a couple more.

from index:

```
app.min.js:19480 Uncaught AssertionError: Expected to have "e3a98be7d2f22205" but got "a25ceb4232891c35" at path "/5/traceId".
```

From trace detail:

```
app.min.js:19480 Uncaught AssertionError: Expected to have "Server Address" but got "sa" at path "/spansBackup/0/binaryAnnotations/2/key".
```

with trace:

```
[{"traceId":"2480ccca8df0fca5","name":"get","id":"2480ccca8df0fca5","timestamp":1457186385375000,"duration":333000,"annotations":[{"timestamp":1457186385375000,"value":"sr","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}},{"timestamp":1457186385708000,"value":"ss","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}}],"binaryAnnotations":[{"key":"http.path","value":"/api/v1/trace/d397ce70f5192a8b","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}},{"key":"srv/finagle.version","value":"6.33.0","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}},{"key":"sa","value":true,"endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}},{"key":"ca","value":true,"endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":57602}}]}]
```

eirslett commented:
Thanks for finding all the bugs! :+1: :+1:

> Expected to have "e3a98be7d2f22205" but got "a25ceb4232891c35"

Are you sure you posted the correct data? I cannot find any of these IDs in that JSON.

eirslett committed:
fix bug where the annotation name wasn't translated to human readable form

eirslett commented:
I fixed the "sa -> Server Address" bug. Also added source map support to karma, so it's easier to see where a bug is coming from.

codefromthecrypt commented:
looks like the gradle test pattern needs to recurse directories:

```
06 03 2016 13:35:51.753:WARN [watcher]: Pattern "/Users/acole/oss/zipkin/zipkin-web/src/main/resources/app/test/*test.js" does not match any file.
```

codefromthecrypt commented:
on the index page, maybe this is an ordering error?

```
Uncaught AssertionError: Expected to have "6ff1c14161f7bde1" but got "9ed44141f679130b" at path "/5/traceId".
```

Here's the  two traces:

```
[{"traceId":"6ff1c14161f7bde1","name":"get","id":"6ff1c14161f7bde1","timestamp":1457186441657000,"duration":4000,"annotations":[{"timestamp":1457186441657000,"value":"sr","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}},{"timestamp":1457186441661000,"value":"ss","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}}],"binaryAnnotations":[{"key":"http.path","value":"/api/v1/trace/2480ccca8df0fca5","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}},{"key":"srv/finagle.version","value":"6.33.0","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}},{"key":"sa","value":true,"endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}},{"key":"ca","value":true,"endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":57603}}]}]
```

```
[{"traceId":"9ed44141f679130b","name":"get","id":"9ed44141f679130b","timestamp":1457186568026000,"duration":4000,"annotations":[{"timestamp":1457186568026000,"value":"sr","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}},{"timestamp":1457186568030000,"value":"ss","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}}],"binaryAnnotations":[{"key":"http.path","value":"/api/v1/trace/2480ccca8df0fca5","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}},{"key":"srv/finagle.version","value":"6.33.0","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}},{"key":"sa","value":true,"endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":9411}},{"key":"ca","value":true,"endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1","port":57603}}]}]
```

---

looks like a slight naming mismatch, too

```
app.min.js:19480 Uncaught AssertionError: Expected to have "Local Address" but got "Local Component" at path "/spansBackup/0/binaryAnnotations/1/key".Assertion.assert @ app.min.js:19480(anonymous function) @ app.min.js:23380ctx.(anonymous function) @ app.min.js:19067success @ app.min.js:56958fire @ app.min.js:4755self.fireWith @ app.min.js:4885done @ app.min.js:10330(anonymous function) @ app.min.js:10696
```

For trace

```
[{"traceId":"f434dda9b8a7c3d8","name":"bootstrap","id":"f434dda9b8a7c3d8","timestamp":1457242094915000,"duration":301000,"annotations":[{"timestamp":1457242094915000,"value":"init","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}},{"timestamp":1457242094915000,"value":"warmup","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}},{"timestamp":1457242095216000,"value":"cr","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}}],"binaryAnnotations":[{"key":"lc","value":"finatra","endpoint":{"serviceName":"zipkin-query","ipv4":"127.0.0.1"}}]}]
```

eirslett committed:
Bugfixes Scala->JavaScript rewrite

1) LC annotations should display as "Local Address"
2) Sort traces; there are implementation differences between browsers,
   some do stable sort, some do unstable sort. When two traces have the
   same duration, we "tie-break" by sorting on trace id too.

eirslett commented:
Thanks a lot for finding all these bugs!
I fixed the "Local Address" bug, I think.
The ordering problem is probably due to implementation differences in browsers; stable vs unstable sort. Both the traces have duration 4 ms/4000 microseconds. V8 does unstable sort, Webkit does stable sort (I think), and Scala appears to do a bit of both, depending on the method used.
I stabilised the sort by "tie-breaking" on trace ID, so if the durations are equal, we sort by trace ID.

eirslett committed:
fix build

codefromthecrypt commented:
@eirslett I agree that the sort order thing isn't a useful signal as indeed any time I get that exception, the traces have the same duration.

LGTM, so merge when ready! (I'm guessing the comparison calls are removed at this point)

eirslett commented:
I'll ditch this PR and make a new one, without the comparisons, and with only a single commit.

eirslett commented:
Closing in favor of #1017.
